src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.Match.Tests.cs

```diff
@@ -945,6 +945,9 @@ private static IEnumerable<(string Pattern, string Input, RegexOptions Options,
                 //mixed lazy and eager counting
                 yield return ("z(a{0,5}|a{0,10}?)", "xyzaaaaaaaaaxyz", options, 0, 15, true, "zaaaaa");
             }
+
+            // Test for a bug in NonBacktracking's subsumption rule for XY subsuming X??Y, which didn't check that X is nullable
```

Comment:
Is there appetite here for adding a bunch more test cases focused on these rules in general? e.g. walk through all the rules adding a few test cases that hit each? We might already have many covered via other tests, just thinking of something a bit more deliberate and thorough. Then we also wouldn't have this comment,, but rather mention the section of tests is broadly those rules.

Reply:
That might be worth it, but I think it would require some thought how to write the tests. Here we figured out separately that it should be checking for an additional condition and the test case was one where that condition didn't hold. For rules that are still correct to the best of our knowledge it might rather be something like having a corpus of "base" regexes and then composing them in various ways that fit the template the rule is looking for. There would have to be also some way of composing inputs that we expect to be matched/not matched. The input generation code we have might be useful for that.

We are also separately thinking about formally proving the correctness of these rules in Lean.